openfl-tutorials/experimental/workflow/FederatedRuntime/101_MNIST/workspace/101_MNIST_FederatedRuntime.ipynb

```diff
@@ -0,0 +1,686 @@
+    "flflow.run()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "50300fed",
+   "metadata": {},
+   "source": [
+    "Let us check the simulation results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a5d77540",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from tabulate import tabulate \n",
+    "\n",
+    "headers = [\"Rounds\", \"Agg Model Validation Score\", \"Local Train loss\", \"Local Model Validation score\"]\n",
+    "print('********** Simulation results **********')\n",
+    "simulation_results = flflow.results\n",
+    "print(tabulate(simulation_results, headers=headers, tablefmt=\"outline\"))\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b5371b6d",
+   "metadata": {},
+   "source": [
+    "### Setup Federation: Director & Envoys"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f270e385",
+   "metadata": {},
+   "source": [
+    "Before we can deploy the experiment, let us create participants in Federation: Director and Envoys. As the Tutorial uses two collaborators we shall launch three participants:\n",
+    "1. Director: The central node in the Federation\n",
+    "2. Portland: The first envoy in the Federation\n",
+    "3. Seattle: The second envoy in the Federation \n",
+    "\n",
+    "The participants can be launched by following steps mentioned in [README]((https://github.com/securefederatedai/openfl/blob/develop/openfl-tutorials/experimental/workflow/FederatedRuntime/101_MNIST/README.md))\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f9d556d0",
+   "metadata": {},
+   "source": [
+    "### Deploy: FederatedRuntime"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5ffd73b6",
+   "metadata": {},
+   "source": [
+    "We now import and instantiate `FederatedRuntime` to enable deployment of experiment on distributed infrastructure. Initializing the `FederatedRuntime` requires following inputs to be provided by the user:\n",
+    "\n",
+    "- `director_info` – director information including fqdn of the director node, port, and certificate information\n",
+    "- `collaborators` - names of the collaborators participating in experiment\n",
+    "- `notebook_path`- path to this jupyter notebook\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1715a373",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "\n",
+    "from openfl.experimental.workflow.runtime import FederatedRuntime\n",
+    "\n",
+    "director_info = {\n",
+    "    'director_node_fqdn':'localhost',\n",
+    "    'director_port':50050,\n",
+    "    'cert_chain': None,\n",
+    "    'api_cert': None,\n",
+    "    'api_private_key': None,\n",
+    "}\n",
+    "\n",
+    "federated_runtime = FederatedRuntime(\n",
+    "    collaborators=collaborator_names,\n",
+    "    director=director_info, \n",
+    "    notebook_path='./101_MNIST_FederatedRuntime.ipynb'\n",
```

Comment:
Just some Friday morning thoughts... Ideally, the experiment owner would only know about the Director, who is their "proxy" to interacting with the compute infrastructure. Now that I think about it again, at this stage it also seems ok for the experiment owner to know about the collaborators that participate. On the other hand, I think the Experiment Owner should not know (or as little as possible) about the Envoys, as they are infrastructure exclusively managed by the Director.

Reply:
In the current implementation collaborator names that are provided to FederatedRuntime become the list of collaborators on which experiment should run. This serves two purposes  
a) Experiment owners get the flexibility to run the experiment even on a subset of participants 
b) Enable Vertical Federated Learning use cases with custom tasks per collaboroator. For e.g. openfl-workspace/experimental/workflow/AggregatorBasedWorkflow/vertical_fl/src/workflow_interface_vertical_fl.py. 
 
Further enhancements can be discussed as part of Runtime refactoring